rfcs/0000-policy_target_scope.md

```diff
@@ -0,0 +1,285 @@
+specific `HTTPRouteRule` by duplicating its `HTTPRouteMatch`es.
+
+Any Kuadrant `Policy` will apply to the entirety of the traffic that is routed through the `HTTPRoute`
+it is eventually attached to. A `Policy` can be attached directly to an `HTTPRoute`. It will apply its
+specification according to the `override` and `default` section of the `Policy`'s `spec`. Let's use a
+simple example:
+
+```yaml
+kind: FakeRLPolicy
+spec:
+  override:
+    enabled: true
+  default:
+    enabled: false
+    rps: 42
+  targetRef:
+    kind: Gateway
+    name: my-gw
+```
+
+This `FakeRLPolicy` defines two values for the `enabled` configuration. Since there is no other
+`FakeRLPolicy` attached to the `my-gw` `Gateway`, its namespace and so forth, see [the specification
+for the hierarchy](https://gateway-api.sigs.k8s.io/references/policy-attachment/#hierarchy), the
+`override` value for `enabled` precedes and as such the `enabled` flag will be `true`. The `rps` value
+will be `42` from the `default` section, as it is omitted in the `override`s. The resulting policy
+(modulo the `targetRef`) to be applied would look like this:
+
+```yaml
+enabled: true
+rps: 42
+```
+
+>**Note:** that the `default`'s `enabled` being specified here is useless. We're just using it as an
+>example to introduce this concepts.
+
+That policy tho has no traffic to police yet. Unless an `HTTPRoute` binds that `Gateway` to a
+`Backend`, no traffic can exist and as such can't be applied to by the policy. As such Kuadrant will
+_not_ create any further of its custom resources (`CR`) until that happens.
+
+Say, we now have a `Backend` (e.g. `foo`) wired to the `my-gw` using some `HTTPRoute` (e.g.
+`foo-route`), none of which have yet any `FakeRLPolicy` attached. Yet, since the `Gateway` used as a
+policy defined, Kuadrant will create a `RateLimit` `CR` and eventually configure `Limitador` to apply
+a rate limit of 42 requests per second, for traffic routed to that `foo` `Backend`. The `RateLimit`
+will be derived from the resulting policy as described above: rate limit enabled at a 42 rps. Which
+traffic will be rate limited, will be matching the union of all the  `HTTPRouteMatch` of each
+`HTTPRouteRule` as defined in that `HTTPRoute`. 
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: foo-route
+spec:
+  parentRefs:
+  - name: my-gw
+  hostnames:
+  - "example.com"
+  rules:
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /foo
+    backendRefs:
+    - name: foo
+      port: 8080
+```
+
+With `foo-route` declaring a single hostname (`example.com`) and a single `HTTPRouteMatch` (anything
+starting with `/foo`), the rate limiting will apply to the same traffic: `example.com/foo*`, so that
+`foo` will not see more than 42 rps on port `8080`.
+
+Given the `override` on the `my-gw` `FakeRLPolicy` there no way for our route to disable rate limiting
```

Comment:
But it would "just" be up to the GW to express this properly, i.e. using a `default` rather than an `override`. Also, there might be a need to say `10` (i.e. less than 42), say when being DDoS'ed, so that'd be an override... The fact that the API lets users express what we think are "none sensible" things, doesn't invalidate it. 

But I think we could also have Kuadrant "merge" the policies based of the most restrictive wins and "only" support `defaults` initially.

Reply:
> Also, there might be a need to say 10 (i.e. less than 42), say when being DDoS'ed, so that'd be an override...

That is a valid use case and it is covered by "most restrictive policy wins"

> But I think we could also have Kuadrant "merge" the policies based of the most restrictive wins 

That is what I call "conditional defaults/overrides" which is worth elaborating/investigating more, but I find it as a complex API to expose to the customers. 

> "only" support defaults initially.

Is this a valid use case? I mean, a gateway provider saying I want (default) 10 rps but I do not care if you as API provider sets 1000 rps. My point being, if the gateway provider applies 10 rps, that 10 is not a magical number it is something specific for the existing resources and constraints.

Reply:
We probably need more real world use cases to be sure.  A clear use case is a gateway provider/infra provider setting a global limit for the gateway to protect it lets say 1000 RPS and wanting that to be a "override" limit. This is a use case we have in other services where we have used limitador to do this at the gateway level as the infra provider in order to enforce a limit based the tier. 
So then the question and use case comes for a default RL. This is less clear as @eguzki points out. That said, I can imagine as a "gateway admin", I don't want any one service to be able to just accidentally consume this entire limit without clearly expressing this intention. I don't want any endpoint deployed to my gateway that some random client can accidentally get into a loop and swamp the gateway. So I now set a sane but restrictive default based on remote ip or some common property to a low number lets say 50 RPS. Now I know if an endpoint is created without any RLP, it can only at worst consume 50 RPS. 
So now it is up to the service provider/developer to say explicitly "I need more than 50 RPS  I need 200 ~RPS" for my service. So the piece I see as missing here is setting an "upper limit" override. IE by default its 100 but you can override it up to 250 but beyond that we need to have a serious conversation. 
@eguzki Does the above make sense in your world? I don't claim to have all the correct use cases here but the above seems sane to me. Interested in your thoughts on the default and being able to set a "upper limit" override.
I think it is also important that we don't think of RLP as the only control involved. As a gateway admin, depending on the environment, I am going to want to setup monitoring so that I can identify what limits are set, and what traffic is flowing to certain endpoints. This will allow me to identify a potential problem before it happens. Also I might have guidelines in place for the teams using my gateway and perhaps I even have an SRE member reviewing any new limits set as part of a git ops flow etc. I think we need to think of RLP as part of a whole rather than the whole.